### PR TITLE
LLM as a Judge Workflow Enhancement

### LINKED ISSUE #397: Evaluation Workflow Enhancement
### What feature do you want to see added?

#392 has the report under the CI summary itself.

We need to enhance this CI to ensure it will comment on the result.

- [x] Only use one comment to avoid adding comments with every change
- [x] In the comment should be similar to the CI summary. In addition, add the commit hash in the comment
- [x] Failing the CI based on our threshold, it will be better if we have a script to check it and read from an environment variable. I'll set up the number in the project environment variable (temporarily, we can include it in the CI script - yaml).

### Upstream changes

#392 

### Are you interested in contributing this feature?

_No response_

### LINKED ISSUE #415: Fix CI due to missing bs4
### Jenkins and plugins versions report

Related broken PR: https://github.com/jenkinsci/resources-ai-chatbot-plugin/pull/404

### What Operating System are you using (both controller, and any agents involved in the problem)?

-

### Reproduction steps

See https://github.com/jenkinsci/resources-ai-chatbot-plugin/pull/404

### Expected Results

CI should be successful

### Actual Results

CI is failing

### Anything else?

_No response_

### Are you interested in contributing a fix?

_No response_

### PR DESCRIPTION
**Pull request by arjun-713**
This PR enhances the existing LLM as a Judge workflow.
It adds a comment similar to the existing workflow summary, if a comment already exists it updates it rather than commenting repeatedly.
It also adds commit SHA to it.

The metric threshold hasnt been modified yet, 0.85 is the preferred threshold but still awaiting mentors approval.

Resolves #397 

Closes #415

### CODE REVIEW

**Comment by arjun-713** [comment]
@berviantoleo Can you please add `eval` label for this as well. Thank you.

**Comment by arjun-713** [comment]
@berviantoleo Due to Github Permission issues commenting on PRs from forks cannot be done as `on: PullRequest`. So had to split it into a separate workflow run. And workflow runs cannot be run unless it is merged. So we are unable to see the commenting on this PR.

File: .github/workflows/eval-comment.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    workflows:
+      - LLM-as-a-Judge
+    types:
+      - completed
+
+permissions:
+  actions: read
+  contents: read
+  issues: write
+  pull-requests: read
```
Comment (berviantoleo) [inline .github/workflows/eval-comment.yml]: You may need to change this as write if you want allow to add comment @arjun-713

**Comment by berviantoleo** [comment]
> @berviantoleo Due to Github Permission issues commenting on PRs from forks cannot be done as `on: PullRequest`. So had to split it into a separate workflow run. And workflow runs cannot be run unless it is merged. So we are unable to see the commenting on this PR.

I don't remember if there is a restriction while using `on: pull_request`. I think you only need to specifically use write permission. If not, we can discuss on the next meeting or on Slack.

I think you can set the permissions.

```
permissions:
    pull_request: write
```

File: .github/workflows/eval-comment.yml (hunk 1 above)
Comment (arjun-713) [reply .github/workflows/eval-comment.yml]: I am trying that.

File: .github/workflows/build-retrieval-cache.yml [hunk 2]
```diff
@@ -0,0 +1,109 @@
+name: Build Retrieval Vector Cache
+
+"on":
```
Comment (berviantoleo) [inline .github/workflows/build-retrieval-cache.yml]:
```suggestion
on:
```

File: .github/workflows/eval-comment.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
+name: LLM-as-a-Judge Comment
+
+"on":
```
Comment (berviantoleo) [inline .github/workflows/eval-comment.yml]:
```suggestion
on:
```

File: .github/workflows/eval.yml [hunk 4]
```diff
@@ -0,0 +1,381 @@
+name: LLM-as-a-Judge
+
+"on":
```
Comment (berviantoleo) [inline .github/workflows/eval.yml]:
```suggestion
on:
```

**Comment by arjun-713** [comment]
> > @berviantoleo Due to Github Permission issues commenting on PRs from forks cannot be done as `on: PullRequest`. So had to split it into a separate workflow run. And workflow runs cannot be run unless it is merged. So we are unable to see the commenting on this PR.
> 
> I don't remember if there is a restriction while using `on: pull_request`. I think you only need to specifically use write permission. If not, we can discuss on the next meeting or on Slack.
> 
> I think you can set the permissions.
> 
> ```
> permissions:
>     pull_request: write
> ```

I tried it and it's the same, the on workflow run action files must be present in the main branch already to run the commenting part as it says in the [docs](https://docs.github.com/en/actions/reference/workflows-and-actions/events-that-trigger-workflows#workflow_run:~:text=run%2E%20For%20information%20about%20each%20activity%20type%2C%20see%20Webhook%20events%20and%20payloads%2E%20By%20default%2C%20all%20activity%20types%20trigger%20workflows%20that%20run%20on%20this%20event%2E%20You%20can%20limit%20your%20workflow%20runs%20to%20specific%20activity%20types%20using%20the%20types%20keyword%2E%20For%20more%20information%2C%20see%20Workflow%20syntax%20for%20GitHub%20Actions%2E-,This%20event%20will%20only%20trigger%20a%20workflow%20run%20if%20the%20workflow%20file%20exists%20on%20the%20default%20branch). So, once it is merged it can work ig.

**Review by cnu1812 — CHANGES_REQUESTED** [review]
I ran through the changes and have a focused list of fixes and improvements that will reduce CI flakiness, make behavior deterministic/consistent, and handle edge cases. Please address the items below; each item includes the file(s) to change and exactly what to do.

1) Sync metric threshold (blocking for correctness)
- Files: chatbot-core/tests/eval/config.json and chatbot-core/tests/eval/metrics.py
- Issue: config.json sets "metric_threshold": 0.85 but metrics.py hard-codes METRIC_THRESHOLD = 0.5.
- Please change metrics.py to derive the threshold from config / environment instead of a hard-coded 0.5. Example approach:
  - Remove METRIC_THRESHOLD constant and add a fallback into build_metrics such as:
    ```py
    import os
    DEFAULT_THRESHOLD = float(os.getenv("METRIC_THRESHOLD", "0.85"))
    def build_metrics(judge_model_name: str, base_url: str, threshold: float = DEFAULT_THRESHOLD):
        ...
    ```
  - Ensure run_evaluation.py passes the configured metric_threshold (from prepare-response-matrix outputs) into build_metrics so the evaluation uses the same 0.85 value that config.json declares.
- Why: keeping config and runtime code in sync avoids surprising test outcomes and wrong pass/fail gating.

2) Use the canonical GitHub token everywhere
- Files: .github/workflows/eval.yml, .github/workflows/eval-comment.yml, and any runner scripts calling GitHub API
- Issue: the workflow sometimes references github.token; callers and scripts should consistently use secrets.GITHUB_TOKEN when authenticating to GitHub REST API.
- Please update workflow env/with blocks and any code that reads the GitHub token to use secrets.GITHUB_TOKEN (and pass it through as GITHUB_TOKEN in env where required). Also ensure scripts expect GITHUB_TOKEN in their environment.
- Why: explicit and consistent secrets.GITHUB_TOKEN usage avoids permission surprises across runner contexts.

3) Robust comment creation/updating (avoid duplicates / handle pagination / transient failures)
- File: chatbot-core/tests/eval/runners/comment_evaluation_report.py
- Issues & changes:
  - Implement pagination when listing PR comments (the current call uses per_page=100 and may miss older comments). Fetch pages until no results or use Link header.
  - Improve existing-comment selection to accept comments authored by the Actions bot robustly (e.g., check user.login == "github-actions[bot]" or accept user.type == "Bot" OR the specific bot login).
  - Add retry/backoff for GitHub API calls and handle 204 No Content responses gracefully.
  - Add a size guard: if the evaluation report is large (e.g., > 60–65 KB), post a short summary comment + attach/link the full report artifact instead of trying to post the whole report text.
- Why: prevents duplicate comments, handles repos with many comments, and avoids hitting API limits / posting truncated content.

4) Harden GitHub API helper
- File: chatbot-core/tests/eval/runners/comment_evaluation_report.py (call_github)
- Issue: uses urllib without retry/backoff and assumes JSON body returned.
- Change: switch to requests.Session with Retry (urllib3 Retry) or implement manual retry + exponential backoff; treat 204 as success and catch JSON decode failures; provide clear error messages when PATCH/POST fails.
- Why: network/API transient issues are common in CI; retries improve reliability.

5) Only reuse retrieval cache artifacts from successful runs
- File: chatbot-core/tests/eval/runners/get_retrieval_cache_run.py
- Issue: artifact selection currently ignores the workflow run status/conclusion.
- Change: when resolving an artifact, validate that artifact["expired"] is False and workflow_run["status"] == "completed" and workflow_run["conclusion"] == "success" before returning run_id. Make error messages explicit if none found.
- Why: prevents reusing partially completed/failed runs' artifacts.

6) Defer config load & validate CLI args (generate_responses)
- File: chatbot-core/tests/eval/runners/generate_responses.py
- Issue: EVAL_CONFIG is loaded at import time and defaults are baked in; also generation-only mode must require an existing retrieval artifact.
- Change:
  - Defer loading of config.json until main() or allow a --config flag so imports do not have side effects.
  - Validate: when --generation-only is used, require --source-responses (or ensure args.responses is a valid retrieval-complete artifact) and error clearly if missing.
- Why: improves testability and prevents surprising behavior when this module is imported.

7) Ollama & prompt-warm behavior improvements
- Files: chatbot-core/tests/eval/runners/generate_responses.py, scripts/start_ollama_service.sh
- Issues & changes:
  - In generate_responses: ensure error messages reference generation_config.response_model and include the Ollama endpoint for easier debugging; log timings (you already added log_ollama_timings — keep it).
  - In start_ollama_service.sh: capture background PID (echo $! > $RESULTS_DIR/ollama.pid), and on failures print the last N lines of the log. Optionally add a trap to kill the server in local dev mode.
- Why: easier debugging of Ollama availability and model pulls.

8) Retry/backoff & logging for DeepEval retries
- File: chatbot-core/tests/eval/runners/run_evaluation.py
- Issue: you retry incomplete cases but do not implement a backoff or log the specific case IDs being retried.
- Change:
  - Before each retry, sleep with exponential backoff (e.g., sleep 2**attempt seconds up to a cap).
  - Log the list of case IDs being retried (so CI logs show exactly what was retried).
  - Ensure max retry attempts and concurrency are configurable (they already are) and document limits.
- Why: lowers judge server load spikes and makes logs actionable.

9) Better diagnostics when shard files are missing or mismatched
- Files: chatbot-core/tests/eval/runners/aggregate_evaluations.py and chatbot-core/tests/eval/runners/merge_shards.py
- Issues & changes:
  - aggregate_evaluations.collect_shard_summaries: when expected_shards != found, include the list of found summary file paths in the error string.
  - merge_shards: when duplicate outputs are detected, include the originating shard name/response file parent directory to ease tracing who produced the duplicate.
- Why: faster triage when CI shows missing or duplicate shards.

**Comment by arjun-713** [comment]
I am looking at the suggestions and will start working on it.

File: chatbot-core/tests/eval/reference/default_faithfulness_template.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
+"""Reference snapshot of DeepEval's default FaithfulnessTemplate.
+
+This file is not imported by the evaluation pipeline. It is kept only to show
+which upstream DeepEval template the Jenkins-specific template was adapted from.
+"""
+
+# pylint: skip-file
```
Comment (berviantoleo) [inline chatbot-core/tests/eval/reference/default_faithfulness_template.py]: why skipping this file?

Comment (arjun-713) [reply chatbot-core/tests/eval/reference/default_faithfulness_template.py]: That was for reference. I will remove the file itself, its not needed for the CI to run. It pulls it automatically during runs.

**Comment by arjun-713** [comment]
Once this is approved, should I rebase #400 and push clean changes?

**Review by berviantoleo — APPROVED** [review]

**Comment by berviantoleo** [comment]
> Once this is approved, should I rebase #400 and push clean changes?

I don't think so. After mergin this, that PR diff should be updated after we update that branch.

**Review by cnu1812 — APPROVED** [review]
lgtm

**Comment by cnu1812** [comment]
@arjun-713 checks are failing, fix them before merge

**Comment by arjun-713** [comment]
> @arjun-713 checks are failing, fix them before merge

It's #415 and I have fixed it in #400

**Comment by arjun-713** [comment]
The problem was insufficient disk space in the runner and I removed the unused backend dependency and also removing models once they are of complete use.
